File: src/amuse/community/metisse/interface.f90
module metisseInterface
    use iso_c_binding
    use store_stars, only: stars
    use track_support
    use z_support
    implicit none
    type(stars) :: star_system
    real(c_double), allocatable :: mass_array(:)

    contains

    ! standard AMUSE interface functions:
    ! initialize
    ! commit_parameters
    ! commit_particles
    ! cleanup_code

    subroutine assign_commons_amuse()

    end subroutine

    function initialize_code()
        implicit none
        integer :: initialize_code

        ! Need to define this front end for METISSE
        call initialize_front_end("AMUSE")
        initial_Z = -1.0_c_double
        write_output_to_file = .false.

        initialize_code=0
    end function
    
    function recommit_parameters()
        implicit none
        integer :: recommit_parameters
        recommit_parameters=0
    end function
    
    function recommit_particles()
        implicit none
        integer :: number_of_particles
        integer :: error
        integer :: recommit_particles
        deallocate(mass_array)
        error = get_number_of_particles(number_of_particles)
        allocate(mass_array(number_of_particles))
        recommit_particles=0
    end function
  
    function cleanup_code()
        implicit none
        integer :: cleanup_code
        cleanup_code=0
    end function
    
    function commit_parameters()
        implicit none
        integer :: commit_parameters
        integer :: error
        real(c_double) :: zpars(20)

        commit_parameters=0

        ! This will read the tracks - so need to have set the paths before
        call METISSE_zcnsts(initial_Z,zpars,'','', error)
        if (error/=0) return

        call assign_commons_main()
    end function
    
    function commit_particles()
        implicit none
        integer :: commit_particles
        integer :: number_of_particles
        integer :: i
        integer :: error
        real(c_double) :: mass
    
        error = get_number_of_particles(number_of_particles)

        allocate(mass_array(number_of_particles))
    
        commit_particles=0
    end function

    ! setters / getters for tracks
    ! metallicity_dir (string)
    ! metallicity_dir_he (string)
    ! z_accuracy_limit (real)
    ! mass_accuracy_limit (real)

    function set_metallicity_dir(metallicity_dir_in)
        implicit none
        character(len=256) :: metallicity_dir_in
        integer :: set_metallicity_dir
        METALLICITY_DIR = metallicity_dir_in
        set_metallicity_dir = 0
    end function

    function get_metallicity_dir(metallicity_dir_out)
        implicit none
        character(len=256) :: metallicity_dir_out
        integer :: get_metallicity_dir
        metallicity_dir_out = METALLICITY_DIR
        get_metallicity_dir = 0
    end function

    function set_metallicity_dir_he(metallicity_dir_he_in)
        implicit none
        character(len=256) :: metallicity_dir_he_in
        integer :: set_metallicity_dir_he
        METALLICITY_DIR_HE = metallicity_dir_he_in
        set_metallicity_dir_he = 0
    end function

    function get_metallicity_dir_he(metallicity_dir_he_out)
        implicit none
        character(len=256) :: metallicity_dir_he_out
        integer :: get_metallicity_dir_he
        metallicity_dir_he_out = METALLICITY_DIR_HE
        get_metallicity_dir_he = 0
    end function

    function set_z_accuracy_limit(z_accuracy_limit_in)
        implicit none
        real(c_double) :: z_accuracy_limit_in
        integer :: set_z_accuracy_limit
        z_accuracy_limit = z_accuracy_limit_in
        set_z_accuracy_limit = 0
    end function

    function get_z_accuracy_limit(z_accuracy_limit_out)
        implicit none
        real(c_double) :: z_accuracy_limit_out
        integer :: get_z_accuracy_limit
        z_accuracy_limit_out = z_accuracy_limit
        get_z_accuracy_limit = 0
    end function

    function set_mass_accuracy_limit(mass_accuracy_limit_in)
        implicit none
        real(c_double) :: mass_accuracy_limit_in
        integer :: set_mass_accuracy_limit
        mass_accuracy_limit = mass_accuracy_limit_in
        set_mass_accuracy_limit = 0
    end function

    function get_mass_accuracy_limit(mass_accuracy_limit_out)
        implicit none
        real(c_double) :: mass_accuracy_limit_out
        integer :: get_mass_accuracy_limit
        mass_accuracy_limit_out = mass_accuracy_limit
        get_mass_accuracy_limit = 0
    end function

    ! setters / getters for misc controls
    ! verbose (bool)
    ! construct_postagb_track (bool)

    function set_verbose(verbose_in)
        implicit none
        logical :: verbose_in
        integer :: set_verbose
        verbose = verbose_in
        set_verbose = 0
    end function

    function get_verbose(verbose_out)
        implicit none
        logical :: verbose_out
        integer :: get_verbose
        verbose_out = verbose
        get_verbose = 0
    end function

    function set_construct_postagb_track(construct_postagb_track_in)
        implicit none
        logical :: construct_postagb_track_in
        integer :: set_construct_postagb_track
        construct_postagb_track = construct_postagb_track_in
        set_construct_postagb_track = 0
    end function

    function get_construct_postagb_track(construct_postagb_track_out)
        implicit none
        logical :: construct_postagb_track_out
        integer :: get_construct_postagb_track
        construct_postagb_track_out = construct_postagb_track
        get_construct_postagb_track = 0
    end function

    ! setters/getters for parameters
    ! initial_metallicity(real)
    ! wd_mass_scheme (string, 256)
    ! use_initial_final_mass_relation(bool)
    ! bhns_mass_scheme (string, 256)
    ! max_ns_mass (real)
    ! allow_electron_capture (bool)

    function set_initial_metallicity(initial_metallicity_in)
        implicit none
        real(c_double) :: initial_metallicity_in
        integer :: set_initial_metallicity
        initial_Z = initial_metallicity_in
        set_initial_metallicity = 0
    end function

    function get_initial_metallicity(initial_metallicity_out)
        implicit none
        real(c_double) :: initial_metallicity_out
        integer :: get_initial_metallicity
        initial_metallicity_out = initial_Z
        get_initial_metallicity = 0
    end function

    function set_wd_mass_scheme(wd_mass_scheme_in)
        implicit none
        character(len=256) :: wd_mass_scheme_in
        integer :: set_wd_mass_scheme
        WD_mass_scheme = wd_mass_scheme_in
        set_wd_mass_scheme = 0
    end function

    function get_wd_mass_scheme(wd_mass_scheme_out)
        implicit none
        character(len=256) :: wd_mass_scheme_out
        integer :: get_wd_mass_scheme
        wd_mass_scheme_out = WD_mass_scheme
        get_wd_mass_scheme = 0
    end function

    function set_use_initial_final_mass_relation(use_initial_final_mass_relation_in)
        implicit none
        logical :: use_initial_final_mass_relation_in
        integer :: set_use_initial_final_mass_relation
        use_initial_final_mass_relation = use_initial_final_mass_relation_in
        set_use_initial_final_mass_relation = 0
    end function

    function get_use_initial_final_mass_relation(use_initial_final_mass_relation_out)
        implicit none
        logical :: use_initial_final_mass_relation_out
        integer :: get_use_initial_final_mass_relation
        use_initial_final_mass_relation_out = use_initial_final_mass_relation
        get_use_initial_final_mass_relation = 0
    end function

    function set_bhns_mass_scheme(bhns_mass_scheme_in)
        implicit none
        character(len=256) :: bhns_mass_scheme_in
        integer :: set_bhns_mass_scheme
        BHNS_mass_scheme = bhns_mass_scheme_in
        set_bhns_mass_scheme = 0
    end function

    function get_bhns_mass_scheme(bhns_mass_scheme_out)
        implicit none
        character(len=256) :: bhns_mass_scheme_out
        integer :: get_bhns_mass_scheme
        bhns_mass_scheme_out = BHNS_mass_scheme
        get_bhns_mass_scheme = 0
    end function

    function set_max_ns_mass(max_ns_mass_in)
        implicit none
        real(c_double) :: max_ns_mass_in
        integer :: set_max_ns_mass
        max_NS_mass = max_ns_mass_in
        set_max_ns_mass = 0
    end function

    function get_max_ns_mass(max_ns_mass_out)
        implicit none
        real(c_double) :: max_ns_mass_out
        integer :: get_max_ns_mass
        max_ns_mass_out = max_NS_mass
        get_max_ns_mass = 0
    end function

    function set_allow_electron_capture(allow_electron_capture_in)
        implicit none
        logical :: allow_electron_capture_in
        integer :: set_allow_electron_capture
        allow_electron_capture = allow_electron_capture_in
        set_allow_electron_capture = 0
    end function

    function get_allow_electron_capture(allow_electron_capture_out)
        implicit none
        logical :: allow_electron_capture_out
        integer :: get_allow_electron_capture
        allow_electron_capture_out = allow_electron_capture
        get_allow_electron_capture = 0
    end function

    ! setters/getters for timestep control
    ! pts_1 to pts_3 (real)

    function set_time_step_pts_1(pts_1_in)
        implicit none
        real(c_double) :: pts_1_in
        integer :: set_time_step_pts_1
        pts_1 = pts_1_in
        set_time_step_pts_1 = 0
    end function

    function get_time_step_pts_1(pts_1_out)
        implicit none
        real(c_double) :: pts_1_out
        integer :: get_time_step_pts_1
        pts_1_out = pts_1
        get_time_step_pts_1 = 0
    end function

    function set_time_step_pts_2(pts_2_in)
        implicit none
        real(c_double) :: pts_2_in
        integer :: set_time_step_pts_2
        pts_2 = pts_2_in
        set_time_step_pts_2 = 0
    end function

    function get_time_step_pts_2(pts_2_out)
        implicit none
        real(c_double) :: pts_2_out
        integer :: get_time_step_pts_2
        pts_2_out = pts_2
        get_time_step_pts_2 = 0
    end function

    function set_time_step_pts_3(pts_3_in)
        implicit none
        real(c_double) :: pts_3_in
        integer :: set_time_step_pts_3
        pts_3 = pts_3_in
        set_time_step_pts_3 = 0
    end function

    function get_time_step_pts_3(pts_3_out)
        implicit none
        real(c_double) :: pts_3_out
        integer :: get_time_step_pts_3
        pts_3_out = pts_3
        get_time_step_pts_3 = 0
    end function

    ! particle management:
    ! new_particle, delete_particle

    function new_particle(index_of_the_particle, mass)
        implicit none
        integer, intent(inout) :: index_of_the_particle
        real(c_double), intent(inout) :: mass
        integer :: new_particle
        index_of_the_particle = star_system%new_star(mass)
    end function

    function delete_star(index_of_the_star)
        implicit none
        integer :: index_of_the_star
        integer :: delete_star
        call star_system%remove_star(index_of_the_star)
        delete_star = 0
    end function

    ! evolving stars:
    ! evolve_for, evolve_one_step
    ! evolve_model, evolve_stars
  
    function evolve_for(index_of_the_star, delta_t)
        implicit none
        integer :: index_of_the_star
        integer :: error
        real(c_double) :: delta_t, time_step, age, previous_time_step
        integer :: evolve_for
        evolve_for = 0

        !write(*,*) 'evolving star ', index_of_the_star, ' for ', delta_t
        call star_system%get_time_step(index_of_the_star, previous_time_step, error)
        ! When the previous time step is 0, the star can not be evolved forward anymore, so we stop.
        ! Possibly, the tracks don't advance further than the current age of the star.
        do while (delta_t > 0.0_c_double .and. previous_time_step > 0.0_c_double)
            call star_system%get_time_step(index_of_the_star, time_step, error)
            if (delta_t < time_step) then
                !write(*,*) 'setting time step of star ', index_of_the_star, ' to ', delta_t
                call star_system%set_time_step(index_of_the_star, delta_t, error)
                time_step = delta_t
            end if
            evolve_for = evolve_one_step(index_of_the_star)
            if (evolve_for /= 0) return
            delta_t = delta_t - time_step
            call star_system%get_age(index_of_the_star, age, error)
            !write(*,*) "age, step: ", age, time_step
            previous_time_step = time_step
        end do
    end function
    
    function evolve_one_step(index_of_the_star)
        implicit none
        integer :: index_of_the_star
        integer :: evolve_one_step
        integer :: error
        real(c_double) :: time_step, nuclear_time_scale
        real(c_double) :: mass
        real(c_double) :: age
        type(track), pointer :: t
        evolve_one_step = 0
   
        call star_system%get_time_step(index_of_the_star, time_step, error)
        if (time_step <= 0.0_c_double) then
            evolve_one_step = 1  ! 1: cannot evolve forward anymore
            return
        end if
        call star_system%get_initial_mass(index_of_the_star, mass, error)
        call star_system%get_age(index_of_the_star, age, error)
        call allocate_track(1, mass) ! allocates tarr. mass is ignored...
        t => tarr(1)
        nuclear_time_scale = t % nuc_time
        if (age + time_step > nuclear_time_scale) then
            time_step = nuclear_time_scale - age
            call star_system%set_time_step(index_of_the_star, time_step, error)
            write(*,*) "reaching end of the nuclear time scale, setting time step to: ", time_step
            evolve_one_step = 1
        end if
        call evolv_metisse(mass, age + time_step, error, 1)
        if (error == 0) then
            call star_system%set_mass(index_of_the_star, t % pars % mass, error)
            call star_system%set_age(&
                index_of_the_star,&
                t % pars % age,&  ! METISSE uses Myr internally, we store years
                error)
            call star_system%set_time_step(&
                index_of_the_star,&
                t % pars % dt,&
                error)
            call star_system%set_luminosity(index_of_the_star, t % pars % luminosity, error)
            call star_system%set_temperature(index_of_the_star, t % pars % Teff, error)
            call star_system%set_radius(index_of_the_star, t % pars % radius, error)
            call star_system%set_epoch(index_of_the_star, t % pars % epoch, error)
            call star_system%set_core_mass(index_of_the_star, t % pars % core_mass, error)
            call star_system%set_core_radius(index_of_the_star, t % pars % core_radius, error)
            call star_system%set_stellar_type(index_of_the_star, t % pars % phase, error)
            call star_system%set_co_core_mass(index_of_the_star, t % pars % McCO, error)
            call star_system%set_spin(index_of_the_star, t % pars % bhspin, error)
        end if
        call dealloc_track()
    end function

    function evolve_model(t_end)
        implicit none
        real(c_double) :: t_end
        integer :: evolve_model
        integer :: number_of_particles
        integer :: error
        integer :: i
        real(c_double) :: mass

        call star_system%get_number_of_stars(number_of_particles)
        do i = 1, number_of_particles
            call star_system%get_initial_mass(i, mass, error)
            call allocate_track(1,mass)
            call evolv_metisse(mass, t_end, error, 1)
            if (error /= 0) then
                call star_system%set_age(i, t_end, error)
                call star_system%set_mass(i, mass, error)
            end if
            call dealloc_track()
        end do

        evolve_model = 0
    end function

    function evolve_stars(indices_of_stars, delta_time, error, n)
        implicit none
        integer, intent(in) :: n
        integer :: error
        integer, intent(in), dimension(n) :: indices_of_stars
        real(c_double) :: delta_time
        integer :: evolve_stars
        integer :: i
        integer :: number_of_stars_to_evolve

        number_of_stars_to_evolve = size(indices_of_stars)

        do i = 1, number_of_stars_to_evolve
            error = evolve_for(indices_of_stars(i), delta_time)
            if (error /= 0) then
                evolve_stars = error
                return
            end if
        end do

        evolve_stars = 0
    end function
  
  function get_age(index_of_the_star, age)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: age
    integer :: get_age
    call star_system%get_age(index_of_the_star, age, get_age)
    age = age * 1.0d+06
  end function
  
  function get_luminosity(index_of_the_star, luminosity)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: luminosity
    integer :: get_luminosity
    call star_system%get_luminosity(index_of_the_star, luminosity, get_luminosity)
  end function
  
  function get_mass(index_of_the_star, mass)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: mass
    integer :: get_mass
    call star_system%get_mass(index_of_the_star, mass, get_mass)
  end function

  function get_metallicity(metallicity)
    implicit none
    real(c_double) :: metallicity
    integer :: get_metallicity
    ! what to do here depends on whether metallicity can be set for individual stars or only globally
    get_metallicity=0
  end function

  function get_epoch(index_of_the_star, epoch)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: epoch
    integer :: get_epoch
    call star_system%get_epoch(index_of_the_star, epoch, get_epoch)
  end function

  function get_core_mass(index_of_the_star, core_mass)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: core_mass
    integer :: get_core_mass
    call star_system%get_core_mass(index_of_the_star, core_mass, get_core_mass)
  end function

  function get_core_radius(index_of_the_star, core_radius)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: core_radius
    integer :: get_core_radius
    call star_system%get_core_radius(index_of_the_star, core_radius, get_core_radius)
  end function

  function get_convective_envelope_mass(index_of_the_star, convective_envelope_mass)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: convective_envelope_mass
    integer :: get_convective_envelope_mass
    call star_system%get_convective_envelope_mass(index_of_the_star, convective_envelope_mass, get_convective_envelope_mass)
  end function

  function get_convective_envelope_radius(index_of_the_star, convective_envelope_radius)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: convective_envelope_radius
    integer :: get_convective_envelope_radius
    call star_system%get_convective_envelope_radius(index_of_the_star, convective_envelope_radius, get_convective_envelope_radius)
  end function

  function get_CO_core_mass(index_of_the_star, CO_core_mass)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: CO_core_mass
    integer :: get_CO_core_mass
    call star_system%get_CO_core_mass(index_of_the_star, CO_core_mass, get_CO_core_mass)
  end function

  function get_main_sequence_lifetime(index_of_the_star, main_sequence_lifetime)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: main_sequence_lifetime
    integer :: get_main_sequence_lifetime
    call star_system%get_main_sequence_lifetime(index_of_the_star, main_sequence_lifetime, get_main_sequence_lifetime)
  end function

  function get_number_of_particles(number_of_particles)
    implicit none
    integer :: number_of_particles
    integer :: get_number_of_particles
    call star_system%get_number_of_stars(number_of_particles)
    get_number_of_particles = 0
  end function
  
  function get_radius(index_of_the_star, radius)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: radius
    integer :: get_radius
    call star_system%get_radius(index_of_the_star, radius, get_radius)
  end function
  
  function get_stellar_type(index_of_the_star, stellar_type)
    implicit none
    integer :: index_of_the_star, stellar_type
    integer :: get_stellar_type
    call star_system%get_stellar_type(index_of_the_star, stellar_type, get_stellar_type)
  end function
  
  function get_temperature(index_of_the_star, temperature)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: temperature
    integer :: get_temperature
    call star_system%get_temperature(index_of_the_star, temperature, get_temperature)
  end function
  
  function get_time_step(index_of_the_star, time_step)
    implicit none
    integer :: index_of_the_star
    real(c_double) :: time_step
    integer :: get_time_step
    call star_system%get_time_step(index_of_the_star, time_step, get_time_step)
    time_step = time_step * 1.0d+06
  end function

  function get_initial_mass(index_of_the_star, mass)
    implicit none
    integer :: index_of_the_star
    double precision :: mass
    integer :: get_initial_mass
    call star_system%get_initial_mass(index_of_the_star, mass, get_initial_mass)
  end function
  
  function set_metallicity(metallicity)
    implicit none
    real(c_double) :: metallicity
    integer :: set_metallicity
    set_metallicity=0
  end function
  
  
end module

File: src/amuse/community/metisse/interface.py
"""
Interface for metisse
"""

from amuse.community import (
    CodeInterface,
    LiteratureReferencesMixIn,
    # LegacyFunctionSpecification,
    # legacy_function,
    remote_function,
)
from amuse.community.interface import se
from amuse.datamodel import Particles, ParticlesSubset
from amuse.units import units, constants


# low level interface class
class MetisseInterface(
    CodeInterface,
    se.StellarEvolutionInterface,
    LiteratureReferencesMixIn,
):
    """
    Low level interface for METISSE

    Details in publication:
        .. [#] Agrawal, P. et al., 2020, https://doi.org/10.1093/mnras/staa2264
        .. [#] Agrawal, P. et al., 2023, https://doi.org/10.1093/mnras/stad2334
    """

    use_modules = ["metisseInterface"]

    def __init__(self, **keyword_arguments):
        CodeInterface.__init__(
            self, name_of_the_worker="metisse_worker", **keyword_arguments
        )
        LiteratureReferencesMixIn.__init__(self)
        self.model_time = 0.0 | units.julianyr

    # Remote functions - getters and setters
    # Note that we should maybe use SI units rather than derived (MSun etc), at
    # least while these are not certain to be the same in the code and in
    # AMUSE...
    @remote_function(can_handle_array=True)
    def get_initial_mass(index_of_the_star="i"):
        returns (mass="d" | units.MSun)

    @remote_function(can_handle_array=True)
    def get_epoch(index_of_the_star="i"):
        returns (epoch="d" | units.julianyr)

    @remote_function(can_handle_array=True)
    def get_core_mass(index_of_the_star="i"):
        returns (core_mass="d" | units.MSun)

    @remote_function(can_handle_array=True)
    def get_core_radius(index_of_the_star="i"):
        returns (core_radius="d" | units.RSun)

    @remote_function(can_handle_array=True)
    def get_convective_envelope_mass(index_of_the_star="i"):
        returns (convective_envelope_mass="d" | units.MSun)

    @remote_function(can_handle_array=True)
    def get_convective_envelope_radius(index_of_the_star="i"):
        returns (convective_envelope_radius="d" | units.RSun)

    @remote_function(can_handle_array=True)
    def get_CO_core_mass(index_of_the_star="i"):
        returns (CO_core_mass="d" | units.MSun)

    @remote_function(can_handle_array=True)
    def get_main_sequence_lifetime(index_of_the_star="i"):
        returns (main_sequence_lifetime="d" | units.Myr)

    # getters and setters for tracks
    # metallicity_dir (string)
    # metallicity_dir_he (string)
    # z_accuracy_limit (float)
    # mass_accuracy_limit (float)

    @remote_function
    def get_metallicity_dir():
        returns (metallicity_dir="s")

    @remote_function
    def set_metallicity_dir(metallicity_dir="s"):
        returns ()

    @remote_function
    def get_metallicity_dir_he():
        returns (metallicity_dir_he="s")

    @remote_function
    def set_metallicity_dir_he(metallicity_dir_he="s"):
        returns ()

    @remote_function
    def get_z_accuracy_limit():
        returns (z_accuracy_limit="d")

    @remote_function
    def set_z_accuracy_limit(z_accuracy_limit="d"):
        returns ()

    @remote_function
    def get_mass_accuracy_limit():
        returns (mass_accuracy_limit="d")

    @remote_function
    def set_mass_accuracy_limit(mass_accuracy_limit="d"):
        returns ()

    # getters and setters for miscellaneous controls
    # verbose (bool)
    # construct_postagb_track (bool)

    @remote_function
    def get_verbose():
        returns (verbose="b")

    @remote_function
    def set_verbose(verbose="b"):
        returns ()

    @remote_function
    def get_construct_postagb_track():
        returns (construct_postagb_track="b")

    @remote_function
    def set_construct_postagb_track(construct_postagb_track="b"):
        returns ()

    # getters and setters for parameters
    # initial_metallicity(real)
    # wd_mass_scheme (string, 256)
    # use_initial_final_mass_relation(bool)
    # bhns_mass_scheme (string, 256)
    # max_ns_mass (real)
    # allow_electron_capture (bool)

    @remote_function
    def get_initial_metallicity():
        returns (initial_metallicity="d")

    @remote_function
    def set_initial_metallicity(initial_metallicity="d"):
        returns ()

    @remote_function
    def get_wd_mass_scheme():
        returns (wd_mass_scheme="s")

    @remote_function
    def set_wd_mass_scheme(wd_mass_scheme="s"):
        returns ()

    @remote_function
    def get_use_initial_final_mass_relation():
        returns (use_initial_final_mass_relation="b")

    @remote_function
    def set_use_initial_final_mass_relation(use_initial_final_mass_relation="b"):
        returns ()

    @remote_function
    def get_bhns_mass_scheme():
        returns (bhns_mass_scheme="s")

    @remote_function
    def set_bhns_mass_scheme(bhns_mass_scheme="s"):
        returns ()

    @remote_function
    def get_max_ns_mass():
        returns (max_ns_mass="d")

    @remote_function
    def set_max_ns_mass(max_ns_mass="d"):
        returns ()

    @remote_function
    def get_allow_electron_capture():
        returns (allow_electron_capture="b")

    @remote_function
    def set_allow_electron_capture(allow_electron_capture="b"):
        returns ()

    @remote_function
    def get_time_step_pts_1():
        returns (fractional_time_step_1="d")

    @remote_function
    def set_time_step_pts_1(fractional_time_step_1="d"):
        returns ()

    @remote_function
    def get_time_step_pts_2():
        returns (fractional_time_step_2="d")

    @remote_function
    def set_time_step_pts_2(fractional_time_step_2="d"):
        returns ()

    @remote_function
    def get_time_step_pts_3():
        returns (fractional_time_step_3="d")

    @remote_function
    def set_time_step_pts_3(fractional_time_step_3="d"):
        returns ()


# high level interface class
class Metisse(se.StellarEvolution):
    """
    High level interface for METISSE
    """
    __interface__ = MetisseInterface

    def __init__(self, **options):
        # self.stopping_conditions = StoppingConditions(self)
        # self.stopping_conditions.supernova_detection = code.StoppingCondition('supernova_detection')
        se.StellarEvolution.__init__(self, MetisseInterface(**options), **options)

    # the definition of the state model of the code
    def define_state(self, handler):
        # for example:
        # handler.set_initial_state("UNINITIALIZED")
        # handler.add_transition("!UNINITIALIZED!STOPPED", "END", "cleanup_code")
        # handler.add_transition("END", "STOPPED", "stop", False)
        # handler.add_transition(
        #     "UNINITIALIZED", "INITIALIZED", "initialize_code")
        # handler.add_method("STOPPED", "stop")
        pass

    # the definition of any properties
    def define_properties(self, handler):
        # handler.add_property("name_of_the_getter", public_name="name_of_the_property")
        pass

    # the definition of the parameters
    def define_parameters(self, handler):
        # handler.add_method_parameter(
        #     "name_of_the_getter",
        #     "name_of_the_setter",
        #     "parameter_name",
        #     "description",
        #     default_value = <default value>
        # )

        

        # Track parameters
        handler.add_method_parameter(
            "get_metallicity_dir",
            "set_metallicity_dir",
            "metallicity_dir",
            "Location of the tracks",
            default_value="./",
            must_set_before_get=True,
        )

        handler.add_method_parameter(
            "get_metallicity_dir_he",
            "set_metallicity_dir_he",
            "metallicity_dir_he",
            "Location of the He tracks",
            default_value="./",
            must_set_before_get=True,
        )

        handler.add_method_parameter(
            "get_z_accuracy_limit",
            "set_z_accuracy_limit",
            "z_accuracy_limit",
            "Metallicity accuracy limit",
            default_value=1.0e-2,
            must_set_before_get=True,
        )

        handler.add_method_parameter(
            "get_mass_accuracy_limit",
            "set_mass_accuracy_limit",
            "mass_accuracy_limit",
            "Mass accuracy limit",
            default_value=1.0e-4,
            must_set_before_get=True,
        )

        # handlers for parameters:
        # initial_metallicity
        # wd_mass_scheme
        # use_initial_final_mass_relation
        # bhns_mass_scheme
        # max_ns_mass
        # allow_electron_capture

        handler.add_method_parameter(
            "get_initial_metallicity",
            "set_initial_metallicity",
            "initial_metallicity",
            "Initial metallicity",
            default_value=-1.0,
            must_set_before_get=True,
        )

        handler.add_method_parameter(
            "get_wd_mass_scheme",
            "set_wd_mass_scheme",
            "wd_mass_scheme",
            (
                "White Dwarf (WD) luminosity calculation method:\n"
                "(1) \"Mestel\" - Shapiro S. L., Teukolsky S. A., 1983\n"
                "(2) \"Modified_mestel\" - Hurley J. R., Shara M. M., 2003"
            ),
            default_value="Modified_mestel",
            must_set_before_get=True,
        )

        handler.add_method_parameter(
            "get_use_initial_final_mass_relation",
            "set_use_initial_final_mass_relation",
            "use_initial_final_mass_relation",
            (
                "If True use the initial final mass relation for white dwarfs "
                "from Han, Z., Posialowski, P., Eggleton, P. P., 1995."
            ),
            default_value=False,
            must_set_before_get=True,
        )

        handler.add_method_parameter(
            "get_bhns_mass_scheme",
            "set_bhns_mass_scheme",
            "bhns_mass_scheme",
            (
                "Neutron Star/Black Hole (NS/BH) type and mass calculation method:\n"
                "(1) \"original_SSE\" - Hurley et al. 2000\n"
                "(2) \"Belczynski2002\" - Belczynski et al. 2002\n"
                "(3) \"Belczynski2008\" - Belczynski et al. 2008\n"
                "(4) \"Eldridge_Tout2004\" - Eldridge J. J., Tout C. A., 2004"
            ),
            default_value="Belczynski2008",
            must_set_before_get=True,
        )

        handler.add_method_parameter(
            "get_time_step_pts_1",
            "set_time_step_pts_1",
            "fractional_time_step_1",
            "Determine timestep for 95% of MS, and HeMS",
            default_value=0.05,
            must_set_before_get=True,
        )

        handler.add_method_parameter(
            "get_time_step_pts_2",
            "set_time_step_pts_2",
            "fractional_time_step_2",
            "Determine timestep for last 5% of MS, cHeBurn, HeHG, and HeGB",
            default_value=0.01,
            must_set_before_get=True,
        )

        handler.add_method_parameter(
            "get_time_step_pts_3",
            "set_time_step_pts_3",
            "fractional_time_step_3",
            "Determine timestep for HG, RGB, EAGB, and TPAGB",
            default_value=0.02,
            must_set_before_get=True,
        )

    def define_particle_sets(self, handler):
        handler.define_set("particles", "index_of_the_star")
        handler.set_new("particles", "new_particle")
        handler.set_delete("particles", "delete_star")

        handler.add_getter("particles", "get_mass", names=("mass",))
        handler.add_getter("particles", "get_radius", names=("radius",))
        handler.add_getter("particles", "get_age", names=("age",))
        handler.add_getter(
            "particles", "get_time_step", names=("time_step",)
        )
        handler.add_getter(
            "particles", "get_temperature", names=("temperature",)
        )
        handler.add_getter(
            "particles", "get_luminosity", names=("luminosity",)
        )
        handler.add_getter(
            "particles", "get_stellar_type", names=("stellar_type",)
        )
        # handler.add_getter("particles", "get_spin", names=("spin",))
        handler.add_getter("particles", "get_epoch", names=("epoch",))
        handler.add_getter(
            "particles",
            "get_main_sequence_lifetime",
            names=("main_sequence_lifetime",),
        )
        handler.add_getter(
            "particles", "get_core_mass", names=("core_mass",)
        )
        handler.add_getter(
            "particles", "get_CO_core_mass", names=("CO_core_mass",)
        )
        handler.add_getter(
            "particles", "get_core_radius", names=("core_radius",)
        )
        handler.add_getter(
            "particles",
            "get_convective_envelope_mass",
            names=("convective_envelope_mass",),
        )
        handler.add_getter(
            "particles",
            "get_convective_envelope_radius",
            names=("convective_envelope_radius",),
        )
        handler.add_getter(
            "particles", "get_initial_mass", names=("initial_mass",)
        )

        handler.add_method("particles", "evolve_one_step")
        handler.add_method("particles", "evolve_for")

    def evolve_model(self, end_time=None, keep_synchronous=True):
        print("evolve_model", end_time, keep_synchronous)
        if not keep_synchronous:
            for particle in self.particles:
                particle.evolve_one_step()
            return

        delta_time = (
            end_time-self.model_time
            if end_time
            else 0.99*min(self.particles.time_step)
        )
        print(f"{delta_time=}")
        for i, particle in enumerate(self.particles):
            print(f"{i} {particle.age} {particle.mass}")
            particle.evolve_for(particle.age + delta_time)
        self.model_time += delta_time
